Sets separated from globals

johnhaddon opened the pull request:
This implements the separating of sets from the scene globals as requested in #1254, making them into first class citizens of the scene. There are several motivations for this :

- Sets have specific semantics and are required to be kept valid with respect to the rest of the scene. Giving them their own compute makes the implementation of this easier.
- Recomputing large sets can be an expensive operation, and we don't want to pay for it when just computing the globals, which should be a quick operation.
- IE has an instancing workflow which generates absolutely huge sets which are totally unused but continue to be generated regardless. By computing sets individually only when needed we hope to avoid this overhead.

This is achieved by adding a `setNamesPlug()` and a `setPlug()` to `ScenePlug`, and adding corresponding hash and compute methods to `SceneNode` to enable their computation. Client code may pull on `setNamesPlug()` to retrieve the names of available sets, and then pull on the `setPlug()` to retrieve an individual set, specifying the desired set using a `"scene:setName"` context variable. A `ScenePlug::set( setName )` method wraps up the latter operation for convenience.

johnhaddon commented:
I did rebase everything so the commits come in as logical order as possible, but I see GitHub has done the usual thing of putting them in the wrong order. https://github.com/johnhaddon/gaffer/commits/separateSetsRebased seems to show them in the right order if that helps with reviewing at all.

johnhaddon commented:
I've added the newline, and also fixed a build problem on OS X.

davidsminor commented:
Actually, forgot to say, there's no new line at the end of GafferScene/SceneAlgo.cpp either. That and the Parent (and some other nodes) seem to be doing weird stuff - I'll investigate

johnhaddon commented:
OK - I've fixed SceneAlgo.cpp, and updated my editor preferences so it won't happen again. Let me know how I can help out in looking at the weirdness.

andrewkaufman commented:
needs a rebase

davidsminor commented:
Ok, so after fixing that, there seems to be a bit of a traversal time/memory regression from gaffer 0.12.1.0 in my big test scene:
```
               time/s             mem/gb
0.12.1.0       195 195 197        5.6 5.6 5.6
this           203 202 202        6.4 6.4 6.3
```
(although I haven't actually confirmed it's generating the same scene yet). I wonder what's going on there...

davidsminor commented:
ah - still broken it looks like.... ignore those numbers till I actually fix it...

davidsminor commented:
dammit - not broken, I was broken - that "sets" plug on the SceneReader was messing with me. Here are the tests again, with scenes that generate identical images when rendered:
```
               time/s             mem/gb
0.12.1.0       194 195 195        5.6 5.6 5.6
this           214 204 205        6.2 6.3 6.3
```
so yeah - same performance characteristics it looks like.

davidsminor commented:
Hmm - dunno about the memory difference, but the performance difference seems to be in operator new according to vtune. Looks like all the extra context construction and hash cache insertions are having an effect

johnhaddon commented:
Thanks for pinpointing that hash problem - I've pushed a fix along with a new unit test which exposed the problem. Also rebased so it's mergeable again.

Thanks for doing the performance tests as well. The timing change doesn't seem too unexpected given the additional context construction in SetFilter, but I'm surprised at the memory difference. Does your test scene contain the sort of large unused sets that should be benefiting from this change? And do you have updated Caribou/Rendering branches so I could do a bit of testing of my own?

andrewkaufman commented:
I think this needs another rebase now I've merged the ScenePathFilterPath stuff

andrewkaufman commented:
Also `GafferSceneUITest.DocumentationTest` seems to be failing

johnhaddon commented:
I've rebased and fixed the test failure.

davidsminor commented:
Actually you know what? I've just shuffled things around on my end a bit so there's fewer context constructions in IEGafferRendering::SetsFilter and IECaribou::SceneInput/AssetOutput, and now it's actually better than before in terms of traversal time: I got a mean of 191 sec over 3 runs. (seems like gaffer 0.12.1.0 is in the mood for taking 197 sec today)

Incidentally, in my IEGafferRendering::SetsFilter, I'm making sure scene:path and scene:filter:inputScene get removed from the context to take advantage of the hash cache - I see that isn't happening in the GafferScene::SetFilter any more. The SetFilter isn't actually being used in that scene, but if I stop doing that in the IEGafferRendering::SetsFilter, the traversal time goes up to about 220 sec, so it looks like it's worth putting that back in.

As for the memory thing-  yeah, it's a bit of a mystery at the moment... One thing I was thinking about, was the fact that it computes some of the sets multiple times every now and then. The hashes were correct - it looked like different threads were just hitting the same computation at the same time. Maybe it's possible that could push the peak allocation higher than it otherwise would, and the reported resident memory would just stay at that level? The repeated set computations could also result in more memory fragmentation I suppose, as we're allocating more of these big tree structures made of ittle bitty objects. Just guessing at the moment really, I suppose I could run it in one thread and see if that makes a difference.

My caribou/rendering branches are both called "separateSets"

davidsminor commented:
...and if you're curious, here's the python script I was running for my tests:

/home/davidm/git/gaffer/mf2test.py

davidsminor commented:
I just did a single threaded scene walk actually, and started printing out the resident memory at the end using resource.getrusage(resource.RUSAGE_SELF).ru_maxrss. Looks like the single threaded walk gets up to 5gb by the end, whereas the 16 threaded one gets up to 6.6gb. Still looks like it might be something to do with sets getting computed multiple times/memory fragmentation etc then... I dunno, can you think of anything else?

Incidentally, the single threaded traversal took 1413.8 seconds, vs 191 seconds for a 16 threaded traversal. If the scaling was linear, then our multithreaded traversal would be 88 sec innit - I wonder where the discrepancy's coming from...

davidsminor commented:
well, looks like calling malloc_trim() at the end of GafferSceneTest::traverseScene() has no effect on the memory stats...

johnhaddon commented:
Fragmentation from using malloc in multiple threads sounds plausible for the memory usage difference between the single threaded traversal vs the threaded - I've been meaning to try out [jemalloc](http://www.canonware.com/jemalloc/) to see if it helps with that, since its reputed to be significantly better. Either way, we need to cut down on the number of allocations we make anyway, especially for the temporary contexts - I was wanting to take a look at that as part of https://github.com/ImageEngine/gaffer/issues/1256. My hope is that we can create the temporary contexts almost entirely on the stack. I've also been thinking about using a custom allocator for the hash cache, since the use pattern is always "lots of little allocations followed by deallocation of everything" I think we could reduce the real allocations to a few very large ones, and actually never deallocate (just reuse). Today I also came across [stack_alloc](https://howardhinnant.github.io/stack_alloc.html) which I think will have a part to play in this somewhere.

But I don't see why fragmentation would account for a difference between this branch and the master - certainly if your test scene has large unused sets, I would expect a reduction in memory use. Does your test scene have sets of that nature?

I can't seem to find your forks on GitHub to get access to the updated Caribou - am I being a muppet?

For the thread scaling, I think the main thing to be improved is the IECore::LRUCache - even cache hits in that require synchronisation at present, and I've seen that become a bottleneck when under contention from lots of small computations. I don't know if that's what's limiting our scaling to 8x rather than 16x in this case though - it could equally be something less within our control like waiting on IO. More interesting profiling to be done...

davidsminor commented:
Really? Dang. I'll just turn 'em into pull requests then:

https://github.com/ImageEngine/caribou/pull/145
https://github.com/ImageEngine/rendering/pull/191

For the fragmentation, the set allocation/deallocation patterns are different for this branch. The master branch just computes/allocates a bunch of sets at the start, whereas this one will potentially compute the same set 4 or 5 times on different threads then discard all but one of them before inserting it into the cache. The master probably ends up with twice as much data (there are two big sets, ObjectType:MeshPrimitive and LOD:100, only one of which is used), but this one seems to have much messier allocation/deallocation patterns, so it wouldn't surprise me if the fragmentation was a lot worse. It does seem surprising that this would swamp the memory saving, so maybe something weird's going on, but it seems kinda plausible.

With the thread scaling, it didn't *look* like IO/locks were playing a big part... I couldn't see much spin time in vtune (or whatever it calls it), and it looked like the cpus were maxed out most of the time. At least part of it must come from doing all these big chunky set computations multiple times in parallel then discarding them, but I don't know if that accounts for all of it.

johnhaddon commented:
Thanks - if I navigate via those links I can actually get to your forks now - the forks just weren't listed on your GitHub main page.

Ah, I was thinking the master branch might end up computing the globals in parallel a few times as well - but I guess that's not the case because they get computed upfront before the traversal in your test script? I'm a bit deep in the UI refactoring right now, but how about I take a look at preventing these parallel computations of the same thing before we think about merging this branch?

davidsminor commented:
Actually, it looks like what happens on the master, is it computes the globals in the constructor for the top level SceneProcedural, and this happens before all the concurrency starts, so it never hits that kind of situation.

johnhaddon commented:
Do you have a before/after memory comparison from a single threaded traversal? That might give us some clues.

davidsminor commented:
Yeah, I can give that a whirl

johnhaddon commented:
I've started looking at this one again. From what I can tell from your Caribou branch, LiveScene will still load all the sets upfront - is that right or have I misunderstood? That might partially explain why the memory savings aren't what we'd expected.

davidsminor commented:
Yeah, in maya it'll be loading sets up front fo sho. As far as I can see it'll totally fall back on the SceneReader behaviour outside maya though (when m_rootScene is NULL), which is how I've been running my tests.

johnhaddon commented:
I've done some testing of my own today, using the same heavy .psc scene I was using for the other sets optimisations. These results are from just plunging straight into `GafferSceneTest.traverseScene()` :

        |0.12.1.0 | This PR
    ---|-----------|------------
6 threads |  230s, 3.15G | 211s, 2.6G
1 thread | 1115s, 2.67G | 1111s, 2.24G

And these are from forcing a computation of all globals/sets before plunging in :

        |0.12.1.0 | This PR
    ---|-----------|------------
6 threads |  214s, 3.76G | 211s, 3.4G
1 thread | 1046s, 2.9G | 1055s, 2.59G

The timings I was getting between runs were a little inconsistent, presumably due to network load, but the memory usage was pretty stable. So for this test case at least, I'd cautiously say that performance is improved, but I'd confidently say that memory usage is reduced across the board. That's on top of the obvious "the UI doesn't lock up computing unneeded globals" benefits...

My test script is at `/home/john/dev/gaffer/bits/pscTest.py` if you want to take a look.

I did have a look at the "stop two threads computing the same thing, just have one wait" idea, and although I'd love to take it on, I think it's too far into the LRUCache internals to be a good one to attack right now.

Out of interest, I'm going to have a go with jemalloc next, to see what effect that has.

johnhaddon commented:
Turns out jemalloc is quite handy :

        | This PR|
    ---|-----------|
malloc, 6 threads |  211s, 2.6G
jemalloc, 6 threads | 192s, 2.31G

But I think that really means that we should stop making so many unnecessary little allocations, so it doesn't matter so much which allocator we're using.

If you want to try jemalloc yourself, this should do the trick :

`env LD_PRELOAD=/home/john/jemallocbuild/lib/libjemalloc.so yourCommandHere`

davidsminor commented:
Sharing data across PathMatchers could help quite a bit with this I guess. For avoiding tons of little allocations, do you reckon the tree nodes could be stored in a big vector and the pointers could be replaced by indices or something?

johnhaddon commented:
OK, I've added three new commits. The first adds a multithreaded "get me all the sets" method which I was using for testing and will shortly be using in the SceneInspector. The last two add your nice hashing optimisations - could you check I've got those right?

johnhaddon commented:
Also rebased to it'll actually merge.
